File: src/BlockTree/BlockTree.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using NSec.Cryptography;

namespace TheBlockTree
{
	// blocks are verified
	public sealed class BlockTree
	{
		public Block Root { get; }
		private readonly BlockIndex blockIndex;

		public IEnumerable<(Block Block, UInt32 Level)> TraverseDepthFirst() => blockIndex.TraverseDepthFirst(Root);
		public IEnumerable<(Block Block, UInt32 Level)> TraverseBreadthFirst() => blockIndex.TraverseBreadthFirst(Root);

		public BlockTree(IEnumerable<Block> unverifiedBlocks)
		{
			// build up index of all blocks
			var unverifiedBlockIndex = new BlockIndex();
			foreach (var unverifiedBlock in unverifiedBlocks)
				unverifiedBlockIndex.Add(unverifiedBlock);

			// get the root node
			Root = unverifiedBlockIndex.TryGetChildren(ReadOnlyMemory<Byte>.Empty, out var foundBlocks) ?
				foundBlocks.Single() :
				throw new NoRootBlockException();

			// verify the blocks and build a tree
			var queue = new Queue<Block>(unverifiedBlockIndex.Count / 2);
			queue.Enqueue(Root);
			while (queue.Count != 0)
			{
				var current = queue.Dequeue();
				if (!unverifiedBlockIndex.TryGetChildren(current.Signature, out var children))
					continue;
				foreach (var child in children)
					if (current.VerifyChild(child))
						queue.Enqueue(child);
					else
						throw new InvalidBlocksException(current, child);
			}

			blockIndex = unverifiedBlockIndex; // now verified
		}
		
		public BlockTree(Byte[] rootData, Key key)
		{
			Root = new Block(ReadOnlySpan<Byte>.Empty, rootData, key);
			blockIndex = new BlockIndex();
			blockIndex.Add(Root);
		}

		public Boolean TryAdd(Block parent, Byte[] data, Key key, [NotNullWhen(true)] out Block? child)
		{
			if (!blockIndex.Contains(parent.Signature))
			{
				child = null;
				return false;
			}
			var childBlock = new Block(parent.Signature.Span, data, key);
			blockIndex.Add(childBlock);
			child = childBlock;
			return true;
		}

		public IReadOnlyList<Block> GetAllBlocks() => blockIndex.GetAllBlocks();

		public class NoRootBlockException : Exception {
			internal NoRootBlockException() { }
		}

		public class InvalidBlocksException : Exception
		{
			public Block Parent { get; }
			public Block Child { get; }
			internal InvalidBlocksException(Block parent, Block child)
			{
				Parent = parent;
				Child = child;
			}
		}
	}
}

File: src/typescript/BlockTree/BlockTree.ts
import { BlockIndex } from "./BlockIndex";
import { Block } from "./Block";
import { Queue } from "../Common/Queue";
import { NodeLevel } from "./Traverse";

export class NoRootBlockError extends Error {}

export class InvalidBlocksError extends Error {
	constructor(
		readonly parent: Block,
		readonly child: Block,
	) {
		super();
	}
}

export class BlockTree {
	public readonly root: Block;
	private readonly blockIndex = new BlockIndex();

	private static readonly emptyBytes = new Uint8Array(0);

	constructor(unverifiedBlocks: IterableIterator<Block>);
	constructor(rootData: Uint8Array, keyPair: nacl.SignKeyPair);

	constructor(param1: IterableIterator<Block> | Uint8Array, param2?: nacl.SignKeyPair) {
		if (param2 !== undefined) {
			const rootData = param1 as Uint8Array;
			const keyPair = param2;

			this.root = new Block(BlockTree.emptyBytes, rootData, keyPair);
			this.blockIndex = new BlockIndex();
			this.blockIndex.add(this.root);
		}
		else {
			const unverifiedBlocks = param1 as IterableIterator<Block>;

			// build up index of all blocks
			const unverifiedBlockIndex = new BlockIndex();
			for (let unverifiedBlock of unverifiedBlocks)
				unverifiedBlockIndex.add(unverifiedBlock);

			// get the root node
			const maybeRootBlock = unverifiedBlockIndex.tryGetChildren(BlockTree.emptyBytes);
			if (maybeRootBlock === undefined || maybeRootBlock.length !== 1)
				throw new NoRootBlockError();
			this.root = maybeRootBlock[0];

			// verify the blocks and build a tree
			const queue = new Queue<Block>(); // using array.shirt() as queue pop is slow O(N)
			queue.push(this.root);
			while (!queue.empty()) {
				const current = queue.pop();
				const children = unverifiedBlockIndex.tryGetChildren(current.signature);
				if (children === undefined)
					continue;
				for (let child of children) {
					if (current.verifyChild(child))
						queue.push(child)
					else
						throw new InvalidBlocksError(current, child);
				}
			}

			this.blockIndex = unverifiedBlockIndex; // now verified
		}
	}

	tryAdd(parent: Block, data: Uint8Array, key: nacl.SignKeyPair): Block | undefined {
		if (!this.blockIndex.contains(parent.signature))
			return undefined;

		const childBlock = new Block(parent.signature, data, key);
		this.blockIndex.add(childBlock);
		return childBlock;
	}

	getAllBlocks(): readonly Block[] {
		return this.blockIndex.getAllBlocks();
	}

	traverseDepthFirst(): IterableIterator<NodeLevel<Block>> {
		return this.blockIndex.traverseDepthFirst(this.root);
	}

	traverseBreadthFirst(): IterableIterator<NodeLevel<Block>> {
		return this.blockIndex.traverseBreadthFirst(this.root);
	}
}